File: Patterns/Command Pattern/CSharp/Command/Stereo/StereoOnWithDvdCommand.cs
namespace Command.Stereo
{
    public class StereoOnWithDvdCommand : ICommand
    {
        private Stereo stereo;
        private int preview;

        public StereoOnWithDvdCommand(Stereo stereo)
        {
            preview = Stereo.OFF;
            this.stereo = stereo;
        }

        public void Execute()
        {
            preview = stereo.State;
            stereo.On();
            stereo.SetDvd();
            stereo.SetVolumn(11);
        }

        public void Undo()
        {
            switch (preview)
            {
                case Stereo.CD:
                    stereo.SetCd();
                    break;
                case Stereo.RADIO:
                    stereo.SetRadio();
                    break;
                case Stereo.OFF:
                    stereo.Off();
                    break;
            }
        }
    }
}

File: Patterns/Command Pattern/Cpp/Command/Stereo/StereoOnWithDvdCommand.cpp
#include "../ICommand.cpp"
#include "./Stereo.cpp"

class StereoOnWithDvdCommand : public ICommand {
private:
    Stereo* stereo;
    int preview;

public:
    StereoOnWithDvdCommand(Stereo* stereo) : stereo{stereo}, preview{stereo->OFF} {}

    void execute() override {
        preview = stereo->getState();
        stereo->on();
        stereo->setDvd();
        stereo->setVolumn(11);
    }

    void undo() override {
        if (preview == stereo->CD) stereo->setCd();
        else if (preview == stereo->RADIO) stereo->setRadio();
        else if (preview == stereo->OFF) stereo->off();
    }

    string getCommandName() override {
        return "StereoOnWithDvdCommand";
    }
};